### PR TITLE
SchemaValidator: Adapting to new DBAL mapping of bigint

### LINKED ISSUE #11377: SchemaValidator is complaining about BIGINT with PHP typehint `int`
### Bug Report

|    Q        |   A
|------------ | ------
| BC Break    | no
| Version     | 3.1.0 and DBAL 4.0.1

#### Summary

With this mapping:
```php
#[ORM\Column(type: Types::BIGINT, nullable: true)]
private ?int $foo = null;
```
... Symfony Profiler is reporting a Mapping Error:

> The field 'Bar#foo' has the property type 'int' that differs from the metadata field type 'string' returned by the 'bigint' DBAL type.

Which is *kinda* right according to the docs at https://www.doctrine-project.org/projects/doctrine-dbal/en/4.0/reference/types.html#bigint :

> Values retrieved from the database are always converted to PHP's integer type if they are within PHP's integer range or string if they aren't.

But according to https://github.com/doctrine/orm/issues/11073#issuecomment-1822464292 , a BIGINT should be typehinted as `int`:

> Then before migrating to 4.0.0, you change the type to string|int, ignore the warning, and **change to int after the migration.**

So what is the recommended typehint now?

I think there should be a recommendation in the docs, maybe I can add it to https://github.com/doctrine/dbal/pull/6336

### PR DESCRIPTION
**Pull request by ThomasLandauer**
Closes #11377

### CODE REVIEW

**Comment by greg0ire** [comment]
Have you tried that on a real project?

**Comment by ThomasLandauer** [comment]
* With the current `BigIntType::class => 'string'`, I'm getting the mapping error with:
  ```php
  private ?int $foo = null;
  ```
  But not with:
  ```php
  private string|array|null $foo = null; // i.e. `string` doesn't have to be the only type
  ```
* With the new `BigIntType::class => 'string|int'`, I'm still getting it with:
  ```php
  private ?int $foo = null;
  ```
   But not with any of these:
  ```php
  private string|int|null $foo = null;
  private int|string|null $foo = null; // i.e. order doesn't matter
  private string|int|array|null $foo = null; // i.e. some "bogus" alternative like `array` is still allowed
  ```
=> So this does indeed look like a step forward to me.

File: src/Tools/SchemaValidator.php [hunk 1]
```diff
@@ -53,7 +53,7 @@ class SchemaValidator
      */
     private const BUILTIN_TYPES_MAP = [
         AsciiStringType::class => 'string',
-        BigIntType::class => 'string',
+        BigIntType::class => 'string|int',
```
Comment (greg0ire) [inline src/Tools/SchemaValidator.php:56]: We still support DBAL 3 though… maybe we should keep `string` when DBAL 3 is installed?

Comment (ThomasLandauer) [reply src/Tools/SchemaValidator.php:56]: Yeah, but how can the version of DBAL be detected here?

Comment (greg0ire) [reply src/Tools/SchemaValidator.php:56]: You could test for the existence of `BigIntType::getName()` with a call to `method_exists()`. That method does not exist with DBAL 4.

**Review by derrabus — CHANGES_REQUESTED** [review]
No tests. 🤷🏻‍♂️

File: src/Tools/SchemaValidator.php (hunk 1 above)
Comment (ThomasLandauer) [reply src/Tools/SchemaValidator.php:56]: OK, but how can I do this inside this `const`?

Comment (greg0ire) [reply src/Tools/SchemaValidator.php:56]: I don't think you can, you'd probably have to convert the const to a property.

**Comment by ThomasLandauer** [comment]
In https://github.com/doctrine/orm/pull/11399 I started over (using git CLI ;-) so I'm closing here.

**Pull request closed without merging**